File: stdin_read_number/std_read_number.cpp
#include <iostream>
#include <string>
#include <algorithm>

auto user_number_input() -> int;
auto read_input() -> std::string;
auto play_again() -> bool;

auto main(int argc, char* argv[]) -> int {
    while (true) {
        auto i = user_number_input();

        if (i <= -1) {
            std::cout << "negative\n";
        }
        else if (i <= 5) {
            std::cout << "smol\n";
        }
        else if (i <= 11) {
            std::cout << "bigger\n";
        }
        else {
            std::cout << "biggest\n";
        }

        if (!play_again()) {
            return 0;
        }
    }
}

auto user_number_input() -> int {
    while (true) {
        std::cout << "Enter a number: " << std::flush;

        try {
            return std::stoi(read_input());
        }
        catch (std::invalid_argument const& e) {
            std::cerr << "Could not parse as signed integer, please try again\n";
        }
    }
}

auto play_again() -> bool {
    std::cout << "Play again?: [y/n] " << std::flush;

    while (true) {
        auto s = read_input();

        std::transform(std::begin(s), std::end(s), std::begin(s),
                [](unsigned char c) { return std::tolower(c); });

        if (s == "y" || s == "yes") {
            return true;
        }
        else if (s == "n" || s == "no") {
            return false;
        }
        else {
            std::cerr << "Could not parse, please enter again: [y/n] " << std::flush;
        }
    }
}

auto read_input() -> std::string {
    std::string buffer;
    std::cin >> buffer;

    return buffer;
}

File: stdin_read_number/std_read_number.rs
use std::io;
use std::io::Write;

fn main() {
    loop {
        match user_number_input() {
            i32::MIN..=-1 => println!("negative"),
            0..=5 => println!("smol"),
            6..=11 => println!("bigger"),
            12..=i32::MAX => println!("biggest"),
        }

        if !play_again() {
            return ();
        }
    }
}

fn user_number_input() -> i32 {
    loop {
        print!("Enter a number: ");
        io::stdout().flush().unwrap();

        match read_input().trim().parse::<i32>() {
            Ok(num) => return num,
            Err(_) => eprintln!("Could not parse as signed integer, please try again"),
        }
    }
}

fn play_again() -> bool {
    print!("Play again?: [y/n] ");
    io::stdout().flush().unwrap();

    loop {
        match read_input().trim().to_lowercase().as_str() {
            "y" | "yes" => return true,
            "n" | "no" => return false,
            _ => {
                print!("Could not parse, please enter again: [y/n] ");
                io::stdout().flush().unwrap();
            }
        };
    }
}

fn read_input() -> String {
    let mut buffer = String::new();

    io::stdin()
        .read_line(&mut buffer)
        .expect("Failed to read line from input");

    buffer
}
